### PR TITLE
[Chat] Announce File Attachments in New Messages

### PR DESCRIPTION
**Pull request by jpeng-ms**
# What
Fixed the issue where live messages don't announce file attachments

# Why

A11Y bug fix

# How Tested

there are 3 scenarios:
- message with attachments only and no content
- message with attachments and text content
- message with attachments and image content

# Process & policy checklist

- [X] I have updated the project documentation to reflect my changes if necessary.
- [X] I have read the [CONTRIBUTING](https://github.com/Azure/communication-ui-library/blob/main/CONTRIBUTING.md) documentation.

**Is this a breaking change?**

- [ ] This change causes current functionality to break.

### CODE REVIEW

File: packages/react-components/src/components/ChatMessage/ChatMessageContent.tsx [hunk 1]
```diff
@@ -164,9 +164,12 @@ export const BlockedMessageContent = (props: BlockedMessageContentProps): JSX.El
 };
 
 const extractContentForAllyMessage = (props: ChatMessageContentProps): string => {
-  if (props.message.content) {
+  let attachments = undefined;
+  /* @conditional-compile-remove(attachment-download) */
+  attachments = props.message.attachments;
+  if (props.message.content || attachments) {
```
Comment (jimchou-dev) [inline packages/react-components/src/components/ChatMessage/ChatMessageContent.tsx:170]: don't need a separate variable or conditional compile here
`if (props.message.content || props.message.attachments) {`

File: packages/react-components/src/components/ChatMessage/ChatMessageContent.tsx [hunk 2]
```diff
@@ -180,6 +183,16 @@ const extractContentForAllyMessage = (props: ChatMessageContentProps): string =>
       parsedContent.replaceChild(imageTextNode, child);
     });
 
+    // Inject attachment names for aria.
+    if (attachments) {
```
Comment (jimchou-dev) [inline packages/react-components/src/components/ChatMessage/ChatMessageContent.tsx:187]: condition compile this whole node, and use `props.message.attachments` instead of `attachments`?

File: packages/react-components/src/components/ChatMessage/ChatMessageContent.tsx (hunk 1 above)
Comment (jpeng-ms) [reply packages/react-components/src/components/ChatMessage/ChatMessageContent.tsx:170]: props.message doesn't have attachments in stable, so it won't work

File: packages/react-components/src/components/ChatMessage/ChatMessageContent.tsx (hunk 2 above)
Comment (jpeng-ms) [reply packages/react-components/src/components/ChatMessage/ChatMessageContent.tsx:187]: props.message doesn't have attachments in stable, so it won't work

File: packages/react-components/src/components/ChatMessage/ChatMessageContent.tsx [hunk 3]
```diff
@@ -180,6 +183,16 @@ const extractContentForAllyMessage = (props: ChatMessageContentProps): string =>
...
+      let attachmentList = '';
+      /* @conditional-compile-remove(attachment-download) */
+      attachmentList = attachments.map((attachment) => attachment.name).join(', ');
+      const attachmentTextNode = document.createElement('div');
+      attachmentTextNode.innerHTML = ` ${attachmentList} `;
+      parsedContent.appendChild(attachmentTextNode);
+    }
```
Comment (Leah-Xia-Microsoft) [inline packages/react-components/src/components/ChatMessage/ChatMessageContent.tsx:194]: Should attachments have some prefix message such as what we have in the string "attachmentCardGroupMessage": "The message has {attachmentCount} attachment"?

Because attachment is a bit different than inline images. Inline images are inline with the message content while attachment is at the end and users might not know what the file names are for.

Comment (jpeng-ms) [reply packages/react-components/src/components/ChatMessage/ChatMessageContent.tsx:194]: i am not sure, i feel like this would be more like a spec from a11y team - i.e. what is the preferred way to announce attachments, they could also argue that they want file names and not counts.

File: packages/react-components/src/components/ChatMessage/ChatMessageContent.tsx (hunk 1 above)
Comment (jpeng-ms) [reply packages/react-components/src/components/ChatMessage/ChatMessageContent.tsx:170]: image

File: packages/react-components/src/components/ChatMessage/ChatMessageContent.tsx [hunk 4]
```diff
@@ -164,9 +164,12 @@ export const BlockedMessageContent = (props: BlockedMessageContentProps): JSX.El
...
 
 const extractContentForAllyMessage = (props: ChatMessageContentProps): string => {
-  if (props.message.content) {
+  let attachments = undefined;
+  /* @conditional-compile-remove(attachment-download) */
+  attachments = props.message.attachments;
+  if (props.message.content || attachments) {
     // Replace all <img> tags with 'image' for aria.
-    const parsedContent = DOMPurify.sanitize(props.message.content, {
+    const parsedContent = DOMPurify.sanitize(props.message.content ?? '', {
```
Comment (emlynmac) [inline packages/react-components/src/components/ChatMessage/ChatMessageContent.tsx:172]: Why is this change needed?

Comment (jpeng-ms) [reply packages/react-components/src/components/ChatMessage/ChatMessageContent.tsx:172]: because line 170

Comment (jpeng-ms) [reply packages/react-components/src/components/ChatMessage/ChatMessageContent.tsx:172]: it doesn't, it takes `string | Node`

File: packages/react-components/src/components/ChatMessage/ChatMessageContent.tsx [hunk 5]
```diff
@@ -207,6 +223,33 @@ const messageContentAriaText = (props: ChatMessageContentProps): string | undefi
       });
 };
 
+/* @conditional-compile-remove(attachment-download) */
+const attachmentCardGroupDescription = (props: ChatMessageContentProps): string => {
+  /* @conditional-compile-remove(attachment-download) */
+  const attachments = props.message.attachments;
+  /* @conditional-compile-remove(attachment-download) */
```
Comment (Leah-Xia-Microsoft) [inline packages/react-components/src/components/ChatMessage/ChatMessageContent.tsx:230]: If the whole function is under a CC, we don't need to put CCs within it.

File: packages/react-components/src/components/ChatMessage/ChatMessageContent.tsx [hunk 6]
```diff
@@ -207,6 +223,33 @@ const messageContentAriaText = (props: ChatMessageContentProps): string | undefi
...
+ * @private
+ */
+export const getAttachmentCountLiveMessage = (
+  attachments: AttachmentMetadata[],
+  attachmentCardGroupMessage: string
+): string => {
+  if (attachments.length === 0) {
+    return '';
+  }
+  /* @conditional-compile-remove(attachment-download) */
```
Comment (Leah-Xia-Microsoft) [inline packages/react-components/src/components/ChatMessage/ChatMessageContent.tsx:246]: Same here

File: packages/react-components/src/components/ChatMessage/ChatMessageContent.tsx [hunk 7]
```diff
@@ -20,6 +20,9 @@ import LiveMessage from '../Announcer/LiveMessage';
 /* @conditional-compile-remove(mention) */
 import { defaultOnMentionRender } from './MentionRenderer';
 import DOMPurify from 'dompurify';
+import { _AttachmentDownloadCardsStrings } from '../AttachmentDownloadCards';
```
Comment (Leah-Xia-Microsoft) [inline packages/react-components/src/components/ChatMessage/ChatMessageContent.tsx:23]: Not neccesary related to this PR but for @JoshuaLai  to confirm, do we still need the type of `_AttachmentDownloadCardsStrings` and the function of `useLocaleStringsTrampoline`?

**Review by Leah-Xia-Microsoft — APPROVED** [review]

**Review by Leah-Xia-Microsoft — APPROVED** [review]

**Review by jimchou-dev — APPROVED** [review]
